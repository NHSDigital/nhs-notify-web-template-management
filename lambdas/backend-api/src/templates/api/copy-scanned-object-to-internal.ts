import { z } from 'zod';
import {
  $GuardDutyMalwareScanStatusPassed,
  type GuardDutyMalwareScanStatusPassed,
} from 'nhs-notify-web-template-management-utils';
import type { LetterUploadRepository } from '../infra/letter-upload-repository';

// Full event is GuardDutyScanResultNotificationEvent from aws-lambda package
// Just typing/validating the parts we use
type CopyScannedObjectToInternalLambdaInput = {
  detail: {
    s3ObjectDetails: {
      objectKey: string;
      versionId: string;
    };
    scanResultDetails: {
      scanResultStatus: GuardDutyMalwareScanStatusPassed;
    };
  };
};
const $CopyScannedObjectToInternalLambdaInput: z.ZodType<CopyScannedObjectToInternalLambdaInput> =
  z.object({
    detail: z.object({
      s3ObjectDetails: z.object({
        objectKey: z.string(),
        versionId: z.string(),
      }),
      scanResultDetails: z.object({
        scanResultStatus: $GuardDutyMalwareScanStatusPassed,
      }),
    }),
  });

export const createHandler =
  ({
    letterUploadRepository,
  }: {
    letterUploadRepository: LetterUploadRepository;
  }) =>
  async (event: unknown) => {
    const {
      detail: {
        s3ObjectDetails: { objectKey, versionId },
      },
    } = $CopyScannedObjectToInternalLambdaInput.parse(event);

    await letterUploadRepository.copyFromQuarantineToInternal(
      objectKey,
      versionId
    );
  };
