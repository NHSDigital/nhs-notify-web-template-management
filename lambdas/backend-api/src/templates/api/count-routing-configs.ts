import type { APIGatewayProxyHandler } from 'aws-lambda';
import { apiFailure, apiSuccess } from './responses';
import type { RoutingConfigClient } from '../app/routing-config-client';
import { logger } from 'nhs-notify-web-template-management-utils/logger';

export function createHandler({
  routingConfigClient,
}: {
  routingConfigClient: RoutingConfigClient;
}): APIGatewayProxyHandler {
  return async function handler(event) {
    const { user: userId, clientId } = event.requestContext.authorizer ?? {};

    if (!clientId || !userId) {
      return apiFailure(400, 'Invalid request');
    }

    const user = { userId, clientId };

    const log = logger.child(user);

    const { data, error } = await routingConfigClient.countRoutingConfigs(
      user,
      event.queryStringParameters
    );

    if (error) {
      log
        .child(error.errorMeta)
        .error('Failed to count routing configs', error.actualError);

      return apiFailure(
        error.errorMeta.code,
        error.errorMeta.description,
        error.errorMeta.details
      );
    }

    return apiSuccess(200, data);
  };
}
