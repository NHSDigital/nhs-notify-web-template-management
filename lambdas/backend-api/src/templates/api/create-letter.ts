import type { APIGatewayProxyHandler } from 'aws-lambda';
import { apiFailure, apiSuccess } from './responses';
import type {
  CreateTemplate,
  ITemplateClient,
} from 'nhs-notify-backend-client';
import { getLetterUploadParts } from '../app/get-letter-upload-parts';

export function createHandler({
  templateClient,
}: {
  templateClient: ITemplateClient;
}): APIGatewayProxyHandler {
  return async function (event) {
    const user = event.requestContext.authorizer?.user;

    if (!user) {
      return apiFailure(400, 'Invalid request');
    }

    const base64body = Buffer.from(event.body ?? '', 'base64');

    const contentType =
      event.headers['Content-Type'] ?? event.headers['content-type'] ?? 'none';

    const { error: getLetterPartsError, data: letterParts } =
      getLetterUploadParts(base64body, contentType);

    if (getLetterPartsError) {
      return apiFailure(getLetterPartsError.code, getLetterPartsError.message);
    }

    const { template, pdf, csv } = letterParts;

    const { data: created, error: createTemplateError } =
      await templateClient.createLetterTemplate(
        template as CreateTemplate,
        user,
        pdf,
        csv
      );

    if (createTemplateError) {
      return apiFailure(
        createTemplateError.code,
        createTemplateError.message,
        createTemplateError.details
      );
    }

    return apiSuccess(201, created);
  };
}
