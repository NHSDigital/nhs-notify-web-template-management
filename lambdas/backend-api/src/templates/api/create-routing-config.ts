import type { APIGatewayProxyHandler } from 'aws-lambda';
import { apiFailure, apiSuccess } from './responses';
import type { RoutingConfigClient } from '../app/routing-config-client';
import { logger } from 'nhs-notify-web-template-management-utils/logger';

export function createHandler({
  routingConfigClient,
}: {
  routingConfigClient: RoutingConfigClient;
}): APIGatewayProxyHandler {
  return async function handler(event) {
    const { user: userId, clientId } = event.requestContext.authorizer ?? {};

    if (!clientId || !userId) {
      return apiFailure(400, 'Invalid request');
    }

    const payload = JSON.parse(event.body || '{}');

    const user = {
      clientId,
      userId,
    };

    const log = logger.child(user);

    const { data, error } = await routingConfigClient.createRoutingConfig(
      payload,
      user
    );

    if (error) {
      log
        .child(error.errorMeta)
        .error('Failed to create routing configuration', error.actualError);

      return apiFailure(
        error.errorMeta.code,
        error.errorMeta.description,
        error.errorMeta.details
      );
    }

    return apiSuccess(201, data);
  };
}
