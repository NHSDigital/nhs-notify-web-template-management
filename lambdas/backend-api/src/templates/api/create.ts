import type { APIGatewayProxyHandler } from 'aws-lambda';
import { TemplateClient } from '@backend-api/templates/app/template-client';
import { apiFailure, apiSuccess } from './responses';

export const handler: APIGatewayProxyHandler = async (event) => {
  const user = event.requestContext.authorizer?.user;

  const dto = JSON.parse(event.body || '{}');

  if (!user) {
    return apiFailure(400, 'Invalid request');
  }

  const client = new TemplateClient(user, false);

  const { data, error } = await client.createTemplate(dto);

  if (error) {
    return apiFailure(error.code, error.message, error.details);
  }

  return apiSuccess(201, data);
};
