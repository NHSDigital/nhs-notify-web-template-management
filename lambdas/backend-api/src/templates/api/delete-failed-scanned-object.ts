import { guardDutyEventValidator } from 'nhs-notify-web-template-management-utils';
import type { LetterFileRepository } from '../infra/letter-file-repository';

export const createHandler =
  ({ letterFileRepository }: { letterFileRepository: LetterFileRepository }) =>
  async (event: unknown) => {
    const {
      detail: {
        s3ObjectDetails: { objectKey, versionId },
      },
    } = guardDutyEventValidator('FAILED').parse(event);

    await letterFileRepository.deleteFromQuarantine(objectKey, versionId);
  };
