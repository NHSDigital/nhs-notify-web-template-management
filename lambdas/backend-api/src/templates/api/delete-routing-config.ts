import type { APIGatewayProxyHandler } from 'aws-lambda';
import { apiFailure, noContent } from './responses';
import type { RoutingConfigClient } from '../app/routing-config-client';
import { logger } from 'nhs-notify-web-template-management-utils/logger';

export function createHandler({
  routingConfigClient,
}: {
  routingConfigClient: RoutingConfigClient;
}): APIGatewayProxyHandler {
  return async function (event) {
    const { user: userId, clientId } = event.requestContext.authorizer ?? {};

    const routingConfigId = event.pathParameters?.routingConfigId;

    if (!userId || !routingConfigId || !clientId) {
      return apiFailure(400, 'Invalid request');
    }

    const user = { userId, clientId };
    const log = logger.child(user);

    const { error } = await routingConfigClient.deleteRoutingConfig(
      routingConfigId,
      user
    );

    if (error) {
      log
        .child(error.errorMeta)
        .error('Failed to delete routing config', error.actualError);

      return apiFailure(
        error.errorMeta.code,
        error.errorMeta.description,
        error.errorMeta.details
      );
    }

    return noContent;
  };
}
