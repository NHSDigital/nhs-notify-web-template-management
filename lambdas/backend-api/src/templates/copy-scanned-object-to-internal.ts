import { createHandler } from './api/copy-scanned-object-to-internal';
import { letterFileRepositoryContainer } from './container';

export const handler = createHandler(letterFileRepositoryContainer());
