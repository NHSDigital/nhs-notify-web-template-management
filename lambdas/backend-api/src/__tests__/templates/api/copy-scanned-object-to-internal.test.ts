import type { LetterUploadRepository } from '../../../templates/infra';
import { makeGuardDutyMalwareScanResultNotificationEvent } from 'nhs-notify-web-template-management-test-helper-utils';
import { mock } from 'jest-mock-extended';
import { createHandler } from '../../../templates/api/copy-scanned-object-to-internal';
import { $GuardDutyMalwareScanStatusFailed } from 'nhs-notify-web-template-management-utils';

function setup() {
  const mocks = { letterUploadRepository: mock<LetterUploadRepository>() };
  const handler = createHandler(mocks);
  return { handler, mocks };
}

it('copies the scanned file to the internal s3 bucket', async () => {
  const { handler, mocks } = setup();

  const event = makeGuardDutyMalwareScanResultNotificationEvent({
    detail: {
      s3ObjectDetails: {
        bucketName: 'quarantine-bucket',
        objectKey: 'template.pdf',
        versionId: 's3-version-id',
      },
      scanResultDetails: {
        scanResultStatus: 'NO_THREATS_FOUND',
      },
    },
  });

  await handler(event);

  expect(
    mocks.letterUploadRepository.copyFromQuarantineToInternal
  ).toHaveBeenCalledWith('template.pdf', 's3-version-id');
});

it('errors if the event has no object key', async () => {
  const { handler, mocks } = setup();

  await expect(
    handler({
      detail: {
        s3ObjectDetails: {
          bucketName: 'quarantine-bucket',
          versionId: 's3-version-id',
        },
        scanResultDetails: {
          scanResultStatus: 'NO_THREATS_FOUND',
        },
      },
    })
  ).rejects.toThrowErrorMatchingSnapshot();

  expect(
    mocks.letterUploadRepository.copyFromQuarantineToInternal
  ).not.toHaveBeenCalled();
});

it('errors if the event has no version id', async () => {
  const { handler, mocks } = setup();

  await expect(
    handler({
      detail: {
        s3ObjectDetails: {
          bucketName: 'quarantine-bucket',
          objectKey: 'template.pdf',
        },
        scanResultDetails: {
          scanResultStatus: 'NO_THREATS_FOUND',
        },
      },
    })
  ).rejects.toThrowErrorMatchingSnapshot();

  expect(
    mocks.letterUploadRepository.copyFromQuarantineToInternal
  ).not.toHaveBeenCalled();
});

it('errors if the event has no virus scan status', async () => {
  const { handler, mocks } = setup();

  await expect(
    handler({
      detail: {
        s3ObjectDetails: {
          bucketName: 'quarantine-bucket',
          objectKey: 'template.pdf',
          versionId: 's3-version-id',
        },
        scanResultDetails: {},
      },
    })
  ).rejects.toThrowErrorMatchingSnapshot();

  expect(
    mocks.letterUploadRepository.copyFromQuarantineToInternal
  ).not.toHaveBeenCalled();
});

it.each($GuardDutyMalwareScanStatusFailed.options)(
  'errors if the event has virus scan status %s',
  async (status) => {
    const { handler, mocks } = setup();

    await expect(
      handler(
        makeGuardDutyMalwareScanResultNotificationEvent({
          detail: {
            s3ObjectDetails: {
              bucketName: 'quarantine-bucket',
              objectKey: 'template.pdf',
              versionId: 's3-version-id',
            },
            scanResultDetails: {
              scanResultStatus: status,
            },
          },
        })
      )
    ).rejects.toThrowErrorMatchingSnapshot();

    expect(
      mocks.letterUploadRepository.copyFromQuarantineToInternal
    ).not.toHaveBeenCalled();
  }
);
