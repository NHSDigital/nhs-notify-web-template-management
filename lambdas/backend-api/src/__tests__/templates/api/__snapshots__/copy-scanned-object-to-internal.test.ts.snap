// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`errors if the event has no object key 1`] = `
"[
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "detail",
      "s3ObjectDetails",
      "objectKey"
    ],
    "message": "Required"
  }
]"
`;

exports[`errors if the event has no version id 1`] = `
"[
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "detail",
      "s3ObjectDetails",
      "versionId"
    ],
    "message": "Required"
  }
]"
`;

exports[`errors if the event has no virus scan status 1`] = `
"[
  {
    "expected": "'NO_THREATS_FOUND'",
    "received": "undefined",
    "code": "invalid_type",
    "path": [
      "detail",
      "scanResultDetails",
      "scanResultStatus"
    ],
    "message": "Required"
  }
]"
`;

exports[`errors if the event has virus scan status ACCESS_DENIED 1`] = `
"[
  {
    "received": "ACCESS_DENIED",
    "code": "invalid_enum_value",
    "options": [
      "NO_THREATS_FOUND"
    ],
    "path": [
      "detail",
      "scanResultDetails",
      "scanResultStatus"
    ],
    "message": "Invalid enum value. Expected 'NO_THREATS_FOUND', received 'ACCESS_DENIED'"
  }
]"
`;

exports[`errors if the event has virus scan status FAILED 1`] = `
"[
  {
    "received": "FAILED",
    "code": "invalid_enum_value",
    "options": [
      "NO_THREATS_FOUND"
    ],
    "path": [
      "detail",
      "scanResultDetails",
      "scanResultStatus"
    ],
    "message": "Invalid enum value. Expected 'NO_THREATS_FOUND', received 'FAILED'"
  }
]"
`;

exports[`errors if the event has virus scan status THREATS_FOUND 1`] = `
"[
  {
    "received": "THREATS_FOUND",
    "code": "invalid_enum_value",
    "options": [
      "NO_THREATS_FOUND"
    ],
    "path": [
      "detail",
      "scanResultDetails",
      "scanResultStatus"
    ],
    "message": "Invalid enum value. Expected 'NO_THREATS_FOUND', received 'THREATS_FOUND'"
  }
]"
`;

exports[`errors if the event has virus scan status UNSUPPORTED 1`] = `
"[
  {
    "received": "UNSUPPORTED",
    "code": "invalid_enum_value",
    "options": [
      "NO_THREATS_FOUND"
    ],
    "path": [
      "detail",
      "scanResultDetails",
      "scanResultStatus"
    ],
    "message": "Invalid enum value. Expected 'NO_THREATS_FOUND', received 'UNSUPPORTED'"
  }
]"
`;
